Ex1

diostam opened the pull request:
This is the Jupyter notebook tutorial with my answers. Please let me know of any comments. 
Regards,
Dio

Commit:
first pass of exercise

Commit:
final corrections

jawnsy commented:
Hey Dio,

Sorry about the delay, I missed the notification for this before. This looks great, I'm glad you're getting up to speed on Jupyter and Python/Pandas so quickly!

**Summary Statistics**

_Q1: What is the most common voltage level for substations? For simplicity, ignore stations that list multiple voltage levels._

Comments: In this case it would be 380kV (with 43 substations in this data set), since we're excluding substations with multiple voltage levels (the 220/110kV substations). I'll try to clarify the question so it's clearer next time. I like your approach with `A.loc[A == A.max()]`; I would've just manually inspected the table or sorted values in descending order.

_Q2: How many unique transmission system operators are captured in this dataset? For simplicity, consider co-owned lines to be a separate operator._

Comments: Great! I like that you used the merge operation and that you verified your result using concat/groupby. My only minor comment here is that variable names like `Operators` should be lowercase, per Python's coding conventions. I'd recommend taking a look at [Python's Style Guide (PEP8)](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)

_Q3: How many transmission lines are longer than 10km?_

Comments: Great, I think this is the correct solution :)

_Q4: What are the two most common resistance/km values?_

Comments: Perfect! Great solution.

**Derived Values**

_Q1:  Which line has the largest total resistance?_

Comments: Looks correct :)

_Q2: What are the lines with the 10 largest X/R ratios? 10 smallest?_

Comments: Looks correct as well. This might nto be the most efficient solution since it sorts all entries, but for quick analysis like this, the efficiency isn't all that relevant.

_Q3: Which line has the highest total impedance? (magnitude only)_

Comments: The question might not have been clear enough - I meant the total impedance (in ohms) rather than ohm/km. So this is missing a step where the calculated Z(ohm/km) is multiplied by the line length (converted to km). The solution you gave identifies the line with maximum per-distance impedance but not total impedance.

Commit:
corrections as per Jonathan comments on 13MARCH2018

diostam commented:
Hello Jonathan, 
Thanks for your comments. The questions are totally clear; it was my fault focusing more on finding the solution and less on what the question actually asks. Please see below some comments about the latest changes. 
Regards,
Dio

**Summary statistics**
Q1: Fixed. I am now dropping the rows with a ';' or a ',' in their index

Q2: Fixed. Thank you sharing this guide. I will have a look asap

**Derived Values**
Q2: Do you mean that we do not have to sort the entire table, but only the XtoR ratio column (plus the l_id)? I added this and compared the running times. I suspect that since the differences will be affected by the creation of the new dataframe (XtorR and l_id columns only), we might even find that sorting the entire dataframe is faster.
Another approach would be to use the 'nlargest' method but I couldn't find the equivalent for lowest. 
Do you have in mind any alternative to sorting? 

Q3: Fixed. The question is very clear. I missed the word "total".

jawnsy commented:
Looks great, I think you everything is correct 😄 I like that you did benchmarks of a few different solutions, it's definitely a good practice for code that's in the hot path. The sort_values comment was really just intended as a side remark; I hope you didn't feel obligated to do that comparison. The data sets we usually work with in pandas are small enough that it efficiency doesn't matter a whole lot (pandas data sets are usually small because I think everything has to be able to fit in memory).

I'm relatively new to pandas so I'm not sure whether they have some internal optimizations (e.g. if you do `sort_values().head(1)` then it's possible to optimize that as a call to `.min()` or `.max()`). From an algorithmic standpoint, there's a slight difference between doing a sort/selection (the sort would be at least `O(n log n)` time complexity) and choosing the top k items (which looks like it could be the lower of `O(n)` and `O(k log k)` where k is the number of elements you want to select). Since in most cases k will be smaller than n, this means this should run closer to linear speed than a sort/select.

There's [`nlargest`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.nlargest.html) and [`nsmallest`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.nsmallest.html) that might do what you want. It's also possible to do a sort first, then use [`head(k)`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.head.html) and [`tail(k)`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.tail.html) on the same (already sorted) DataFrame, rather than sorting twice, though pandas might have some optimizations for that pattern as well.

In case you're not familiar, I'm using [Big O notation](https://en.wikipedia.org/wiki/Big_O_notation), which gives an indication of the order of an algorithm - how well it scales in response to the size of the data set (`n`). Smaller-order functions are better, as they are expected to complete in less time. So `O(n log n)` is supralinear/higher-than-linear (not as bad as quadratic but not as good as linear).

Looks like there's some good discussion on StackOverflow about these:

* [Pandas good approach to get top-n records within each group](https://stackoverflow.com/questions/20069009/pandas-good-approach-to-get-top-n-records-within-each-group)
* [Optimal algorithm for returning top k values from an array of length N](https://stackoverflow.com/questions/4956593/optimal-algorithm-for-returning-top-k-values-from-an-array-of-length-n) (Usually these algorithms can be generalizable by changing the comparison, min/max can usually be done the same way)

Anyways, I think we can close this pull request now. The idea was just to have you open this to make it easier to review, and you got all the answers correct 😃 Hopefully you enjoyed the exercise and feel comfortable using Jupyter and pandas now.
